Bluetooth: Host: Invoke tx callbacks when channel is deleted

alwa-nordic opened the pull request:
When bt_l2cap_send_pdu() succeeds, it transfers buffer ownership to the stack, which must eventually invoke the provided callback. This contract is honored in all paths where transmission becomes impossible:

- Normal transmission: callback invoked with err=0 after HCI Number of Completed Packets event (tx_notify_process)
- Send errors (after tx allocated): callback invoked with err=-ESHUTDOWN via conn_tx_destroy
- Send errors (before tx allocated): callback invoked with the specific error code in send_buf error_return path
- Connection disconnect: callbacks invoked with err=-ESHUTDOWN via process_unack_tx -> conn_tx_destroy for all PDUs in tx_pending

However, when a channel is deleted (l2cap_chan_del), PDUs remaining in the tx_queue are dropped without invoking their callbacks, violating the ownership contract.

Fix this by extracting and invoking any non-NULL callbacks from the closure stored in buf->user_data before releasing the buffers. The callback is invoked with err=-ESHUTDOWN, making this path analogous to process_unack_tx: both drain queues of unsent PDUs when transmission becomes impossible due to external events (channel deletion vs connection disconnect). The only difference is the buffer lifecycle stage - in l2cap_chan_del, PDUs are still in tx_queue (closure in buf->user_data), while in process_unack_tx, they've progressed to tx_pending (callback in bt_conn_tx struct).

Note: conn_tx_destroy() cannot be used here because no bt_conn_tx struct has been allocated yet - the closure is still in buf->user_data.

alwa-nordic commented:
The twister failures are unrelated. It seems some python package is missing in CI.

Thalley commented:
> The twister failures are unrelated. It seems some python package is missing in CI.

That issue should be fixed by now - Has this PR been rebased recently?

Thalley commented on subsys/bluetooth/host/l2cap.c:
Would it be possible / make sense for this to call `conn_tx_destroy` in some way? 

It seemingly does the same, and would help ensure that `-ESHUTDOWN` is used in all cases

alwa-nordic replied:
I'm not sure how to structure this.

We cannot use `conn_tx_destroy` as is because we don't have a `bt_conn_tx` yet. But we also shouldn't if we want to keep the layers separate. Both `bt_conn_tx` and `conn_tx_destroy` are internal to `conn`. Maybe I should not have mentioned it in the code comment given that.

I think it's better to think of it like this: It's always `l2cap` that decides the callback should be invoked with `-ESHUTDOWN`. Architecturally, `conn` should destroy its internal `bt_conn_tx` and tell `l2cap` (passing back some l2cap user data) and l2cap should then decide to invoke the higher up callback with `-ESHUTDOWN`. It's just a kind of optimization that we end up passing the upper layer callback to `conn`.

I would like to re-architecture the conn l2cap interface to get rid of the arbitrary callback passed to `conn`. Instead `conn` could just signal l2cap that a particular connection got a num complete and l2cap should know which channel that was and the channel should know what action to take. But I'm not going to do this now.

alwa-nordic commented:
Rebased. Maybe it's needed to fix CI.

MarekPieta reviewed (approved):
Thanks for fixing this one. I confirmed locally that the fix works on my end

alwa-nordic commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -278,6 +278,24 @@ static void l2cap_chan_del(struct bt_l2cap_chan *chan)
+			 *
+			 * We use -ESHUTDOWN which is consistent with process_unack_tx(), which
+			 * handles PDUs that are already submitted to the controller but not yet
+			 * acknowledged.
```
Update pushed https://github.com/zephyrproject-rtos/zephyr/compare/80f49366ba5726603706211caa3ff2557836aaf7..05d2f712f09d7e38a16127f4e9a733d7720bfd66

HaavardRei replied:
Nice, thanks! Perhaps the part at least in public l2cap.h should be split into a separate commit?

alwa-nordic replied:
Separate commit done.

alwa-nordic committed:
Bluetooth: Host: Add documentation for L2CAP tx buffer callbacks

Document the ownership contract and callback semantics for L2CAP
transmission buffers:

- Extend tx_queue documentation in bt_l2cap_le_chan to explain SDU/PDU
  storage, callback handling, and transmission order semantics
- Document bt_l2cap_create_frag responsibilities regarding callback
  invocation after HCI Number of Completed Packets or on disconnect
- Document bt_l2cap_send_pdu callback contract: invoked exactly once
  after transmission completes or with -ESHUTDOWN on disconnect

These documentation additions clarify the requirements for proper
buffer lifecycle management and callback invocation throughout the
L2CAP transmission path.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: Host: Invoke tx callbacks when channel is deleted

When bt_l2cap_send_pdu() succeeds, it transfers buffer ownership to the
stack, which must eventually invoke the provided callback. This contract
is honored in all paths where transmission becomes impossible:

- Normal transmission: callback invoked with err=0 after HCI Number of
  Completed Packets event (tx_notify_process)
- Send errors (after tx allocated): callback invoked with err=-ESHUTDOWN
  via conn_tx_destroy
- Send errors (before tx allocated): callback invoked with the specific
  error code in send_buf error_return path
- Connection disconnect: callbacks invoked with err=-ESHUTDOWN via
  process_unack_tx -> conn_tx_destroy for all PDUs in tx_pending

However, when a channel is deleted (l2cap_chan_del), PDUs remaining in
the tx_queue are dropped without invoking their callbacks, violating the
ownership contract.

Fix this by extracting and invoking any non-NULL callbacks from the
closure stored in buf->user_data before releasing the buffers. The
callback is invoked with err=-ESHUTDOWN, making this path analogous to
process_unack_tx: both drain queues of unsent PDUs when transmission
becomes impossible due to external events (channel deletion vs connection
disconnect). The only difference is the buffer lifecycle stage - in
l2cap_chan_del, PDUs are still in tx_queue (closure in buf->user_data),
while in process_unack_tx, they've progressed to tx_pending (callback in
bt_conn_tx struct).

Note: conn_tx_destroy() cannot be used here because no bt_conn_tx struct
has been allocated yet - the closure is still in buf->user_data.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

KyraLengfeld commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -267,7 +267,30 @@ struct bt_l2cap_le_chan {
+	 * remaining part of an already partially transmitted SDU/PDU due to
+	 * L2CAP segmentation and fragmentation.
+	 *
+	 * This is the outbox for a single channel. Channels may be serviced in
```
Nitpick: While `outbox`  is clear in context for me, I don't think this word is usually used (for queues, or anything really in Zephyr), so might confuse other readers. Feel free to leave it as is though.
